File: sharpen/F-OMP/dosharpen.f90
!  Subroutine to sharpen an image by convolving with a filter function. 
!  The filter is a combination of a Gaussian (to remove noise) and a 
!  Laplacian (to detect the edges). Input and output is via Portable 
!  Grey Map (PGM) files - note that the input file must have a specific
!  header format.
!
!  In this version of the program the image processing is parallelised 
!  using OpenMP threads. The master thread reads in the fuzzy image and
!  stores it in shared memory. Further threads are launched, the convolution
!  computation is distributed over all threads (including the master thread),
!  and the result stored in shared memory. Finally the master thread adds the 
!  convolution result to the fuzzy image and writes the resulting sharp image
!  to file.
!
!  David Henty, EPCC, August 2009
!  OpenMP version: Elena Breitmoser, EPCC, August 2009
!  Minor modifications: Arno Proeme, EPCC, March 2013
!

subroutine dosharpen(filename, nx, ny)

  use omp_lib

  implicit none

  integer, parameter :: d = 8 
  ! Sets the linear range of the sharpen filter as measured from any given pixel:
  ! only pixels within a (2d+1)*(2d+1) square centered on the pixel are used to 
  ! compute its new value.

  

  double precision, parameter :: norm = dble((2*d-1)*(2*d-1))
  double precision, parameter :: scale = 2.0

  double precision :: tstart, tstop, time

  double precision, external ::  filter

  integer, parameter :: maxlen = 32
  character*(maxlen) :: filename
  integer, parameter :: iounit = 10

  integer :: i, j, k, l, nx, ny, xpix, ypix, pixcount, nthreads, threadid

  integer, dimension(nx, ny) :: fuzzy                             ! Will store the fuzzy input image when it is first read in from file
  double precision, dimension(1-d:nx+d, 1-d:ny+d) :: fuzzyPadded  ! Will store the fuzzy input image plus additional border padding
  double precision, dimension(1-d:nx+d, 1-d:ny+d) :: convolution  ! Will store the convolution of the filter with the fuzzy image
  double precision, dimension(nx, ny) :: sharp                    ! Will store the sharpened image obtained by adding the convolution to the fuzzy image

  fuzzy(:,:) = 0
  convolution(:,:) = 0.0D0

  write(*,fmt='(" Using a filter of size", I3," x ",I3," pixels")') 2*d+1, 2*d+1
  write(*,*) 'Reading input image file: ', filename
  flush(6)
  call pgmread(filename, fuzzy, nx, ny, xpix, ypix)  ! Loads fuzzy image from file into fuzzy image array
  write(*,*) '... done'
  write(*,*)
  flush(6)

  if (xpix == 0 .or. ypix == 0 .or. nx /= xpix .or. ny /= ypix) then

     write(*,*) 'Error reading ', filename
     flush(6)
     stop

  end if

  ! Initialise image arrays
  fuzzyPadded(:,:) = 0.0
  fuzzyPadded(1:nx, 1:ny) = fuzzy(1:nx, 1:ny)
  sharp(:,:) = 0.0

  write(*,*) 'Starting calculation ...'
  flush(6)

!$omp parallel
  call printlocation()
!$omp end parallel   
  
  tstart  = omp_get_wtime()

  ! Start of parallel region where filter is applied to fuzzy image
  !$omp parallel private(i, j, pixcount, k, l, threadid)

  nthreads  = omp_get_num_threads()
  threadid = omp_get_thread_num()

  pixcount = 0

  do i = 1, nx
     do j = 1, ny
        pixcount = pixcount + 1

        ! Computation of convolution allocated to threads using simple cyclic distribution
        ! i.e. consecutively numbered threads take turns computing convolution for consecutive pixels
        if (mod(pixcount-1, nthreads) == threadid) then

           do k = -d, d
              do l = -d, d
                 
                 convolution(i,j) = convolution(i,j) + filter(d, k, l)*fuzzyPadded(i+k, j+l)

              end do
           end do
        
        end if

     end do
  end do

  !$omp end parallel 
  !End parallel region and convolution computation

  tstop = omp_get_wtime()
  time  = tstop - tstart

  write(*,*) '.. finished'
  write(*,*)

  ! Add rescaled convolution to fuzzy image to obtain sharp image
  sharp(1:nx, 1:ny) = fuzzy(1:nx, 1:ny) - scale/norm * convolution(1:nx, 1:ny)

  filename = 'sharpened.pgm'
  write(*,*) 'Writing output image file: ', filename
  write(*,*)
  flush(6)

  !  Only save the core of the sharpened image to remove edge effects
  call pgmwrite(filename, sharp(d+1:nx-d,d+1:ny-d), nx-2*d, ny-2*d)
  write(*,*) '... done'
  write(*,*)
  write(*,fmt='(" Calculation time was ", f6.3, " seconds")') time
  flush(6)

end subroutine dosharpen

File: src/C-OMP/dosharpen.c
/*  Function to sharpen an image by convolving with a filter function. The
 *  filter is a combination of a Gaussian (to remove noise) and a Laplacian
 *  (to detect the edges). Input and output is via Portable Grey Map (PGM)
 *  files - note that the input file must have a specific header format.
 *
 *  In this version of the program the image processing is parallelised 
 *  using OpenMP threads. The master thread reads in the fuzzy image and
 *  stores it in shared memory. Further threads are launched, the convolution
 *  computation is distributed over all threads (including the master thread),
 *  and the result stored in shared memory. Finally the master thread adds the 
 *  convolution result to the fuzzy image and writes the resulting sharp image
 *  to file.
 *  
 *  David Henty, EPCC, September 2009
 *  Arno Proeme, EPCC, March 2013 (minor modifications)
 *  Dominic Sloan-Murphy, EPCC, June 2014 (improved consistency with other versions)
 */

#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include "utilities.h"
#include "sharpen.h"

void dosharpen(char *infile, int nx, int ny)
{
  int d = 8;  
  /* Sets the linear range of the sharpen filter as measured from any given pixel:
     only pixels within a (2d+1)*(2d+1) square centered on the pixel are used to 
     compute its new value. */

  double  norm = (2*d-1)*(2*d-1);
  double scale = 2.0;
  
  int nthreads, threadid;
  int xpix, ypix, pixcount;
  
  int i, j, k, l;
  double tstart, tstop, time;
  
  int fuzzy[nx][ny];                   /* Will store the fuzzy input image when it is first read in from file                        */
  double fuzzyPadded[nx+2*d][ny+2*d];  /* Will store the fuzzy input image plus additional border padding                            */
  double convolution[nx][ny];          /* Will store the convolution of the filter with the fuzzy image                              */
  double sharp[nx][ny];                /* Will store the sharpened image obtained by adding the  convolution to the fuzzy image      */
  double sharpCropped[nx-2*d][ny-2*d]; /* Will store the sharpened image cropped to remove a border layer distorted by the algorithm */
  
  char *outfile = "sharpened.pgm";
  
  /* Initialise image arrays */
  for (i=0; i < nx; i++)
    {
      for (j=0; j < ny; j++)
        {
          fuzzy[i][j] = 0;
          sharp[i][j] = 0.0;
        }
    }
  
  printf("Using a filter of size %d x %d\n", 2*d+1, 2*d+1);
  printf("\n");

  printf("Reading image file: %s\n", infile);
  fflush(stdout);
       
  pgmread(infile, fuzzy, nx, ny, &xpix, &ypix);
  printf("... done\n\n");
  fflush(stdout);
  
  if (xpix == 0 || ypix == 0 || nx != xpix || ny != ypix)
    {
      printf("Error reading %s\n", infile);
      fflush(stdout);
      exit(-1);
    }
  
  /* Initialise image array */
  for (i=0; i < nx+2*d; i++)
    {
      for (j=0; j < ny+2*d; j++)
        {
          fuzzyPadded[i][j] = 0.0;
        }
    }
  
  /* Transfer fuzzy image into padded array */
  for (i=0; i < nx; i++)
    { 
      for (j=0; j < ny; j++)
        {
          fuzzyPadded[i+d][j+d] = fuzzy[i][j];
        }
    }
  
  printf("Starting calculation ...\n");

  /* Print out thread location. */
#pragma omp parallel
{
  printlocation();
}  

  tstart = omp_get_wtime();

  /* Start of parallel region where filter is applied to fuzzy image */
#pragma omp parallel private(i, j, k, l, pixcount, threadid)
{
  nthreads  = omp_get_num_threads();
  threadid = omp_get_thread_num();
  
  pixcount = 0;
  
  for (i=0; i < nx; i++)
    {
      for (j=0; j < ny; j++)
        {
          /* Computation of convolution allocated to threads using simple cyclic distribution
             i.e. consecutively numbered threads take turns computing convolution for consecutive pixels */
          if (pixcount%nthreads  == threadid)
            {
              for (k=-d; k <= d; k++)
                {
                  for (l= -d; l <= d; l++)
                    {
                      convolution[i][j] = convolution[i][j] + filter(d,k,l)*fuzzyPadded[i+d+k][j+d+l];
                    }
                }
            }
          pixcount += 1;
        }
    }
}
  /* End of parallel region and convolution computation */
  
  tstop = omp_get_wtime();
  time = tstop - tstart;
  
  printf("... finished\n");
  printf("\n");
  fflush(stdout);
  
  /* Add rescaled convolution to fuzzy image to obtain sharp image */
  for (i=0 ; i < nx; i++)
    {
      for (j=0; j < ny; j++)
        {
          sharp[i][j] = fuzzyPadded[i+d][j+d] - scale/norm * convolution[i][j];
        }
    }
  
  printf("Writing output file: %s\n", outfile);
  printf("\n");
  
  /* Only save the core of the sharpened image to remove edge effects */
  for (i=d ; i < nx-d; i++)
    {
      for (j=d; j < ny-d; j++)
        {
          sharpCropped[i-d][j-d] = sharp[i][j];
        }
    }
  
  pgmwrite(outfile, sharpCropped, nx-2*d, ny-2*d);
  
  printf("... done\n");
  printf("\n");
  printf("Calculation time was %f seconds\n", time);
  fflush(stdout);
}
